services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,191 @@
+
+    /**
+     * Chains of chunked messages, it is filling by messages one by one as soon as the next linked message is received.
+     */
+    private final List<LinkedList<ChunkedInfoMessage>> messagesChains = new ArrayList<>();
+
+    @Autowired
+    private MessageProducer messageProducer;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<InfoMessage> sendAndGet(String topic, Message message) {
+        CompletableFuture<InfoMessage> future = new CompletableFuture<>();
+        pendingRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<List<ChunkedInfoMessage>> sendAndGetChunked(String topic, Message message) {
+        CompletableFuture<List<ChunkedInfoMessage>> future = new CompletableFuture<>();
+        pendingChunkedRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for the pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Received invalid message: {}", message);
+            return;
+        } else if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+            throw new MessageException(error);
+        }
+
+        InfoMessage infoMessage = (InfoMessage) message;
+        if (infoMessage instanceof ChunkedInfoMessage) {
+            ChunkedInfoMessage chunked = (ChunkedInfoMessage) infoMessage;
+            processChunkedMessage(chunked);
+        } else {
+            CompletableFuture<InfoMessage> request = pendingRequests.remove(infoMessage.getCorrelationId());
+            if (request != null) {
+                request.complete(infoMessage);
+            } else {
+                logger.warn("Received non-pending message: {}", infoMessage);
+            }
+        }
+    }
+
+    /**
+     * Tries to find and collect all chunked messages into one chain.
+     */
+    private void processChunkedMessage(final ChunkedInfoMessage received) {
+        ChunkedInfoMessage message = received;
+
+        LinkedList<ChunkedInfoMessage> currentChain = messagesChains.stream()
```

Comment:
You can define a Map and use "next-request-id" of pending chains as key in it. (It will require to update this map when you receive next chunk but even in this case it will be faster - it is one map.remove + map.put operation on each received chunk.

Also this approach will significantly simplify following loop.